packages/edit-site/src/components/list/header.js

```diff
@@ -24,8 +26,8 @@ export default function Header( { templateType } ) {
 				{ postType.labels?.name }
 			</Heading>
 
-			<div>
-				<Button variant="primary">{ postType.labels?.add_new }</Button>
+			<div className="edit-site-list-header__right">
+				<AddNewTemplate templateType={ templateType } />
```

Comment:
Not something to action, but just thinking about the future. Later on I wonder if we can consider an architecture like this:
```jsx
<EntityListing type="postType" name="template">
    <HeaderActions>
        <AddNewTemplate />
    </HeaderActions>
</EntityListing>
<EntityListing type="postType" name="template-part">
    <HeaderActions>
        <AddNewTemplatePart />
    </HeaderActions>
</EntityListing>
```

This way there isn't an individual component that manages so much complexity for each different type.

Also thinking about how we might potentially add more entity types here in the future (menus?), and that this might also end up being useful for posts or pages too. Possibly the start of reactifying some of WordPress admin.

Reply:
Looks interesting! Though I believe we don't need a fancy name Like `EntityListing` for that, to me it's just a `Context.Provider` 😅 .

It's also not ideal to render them all even if they won't be rendered in the end, the hooks will still run and potentially other side effects. I think we can re-think this in #36488 and use a route system for that. They are, just search params in the end.

Reply:
The way hooks execute is more of an implementation detail. The code example is just pseudocode.

The point I'm trying to make is that in the future the architecture can support some level of declarative customisation.

At the moment the solution is to add more implicit `if` statements in components like `AddNewTemplate`, but that should instead be inverted so that the control is in the hands of the implementor of these list pages. Imagine it from the point of view of this being a library of components for a user that might want to build their own list of posts, but with custom 'add' buttons, custom columns.

All food for thought for after 5.9 though, but I think there's a good chance that will be needed—initially for menus, but the future might hold more.

Reply:
100% agree! I also agree that it's a bit early to overthink these at the time. There's something we can do now though, like something you mentioned in https://github.com/WordPress/gutenberg/pull/36379#discussion_r751993057 and https://github.com/WordPress/gutenberg/pull/36379#discussion_r751981614.